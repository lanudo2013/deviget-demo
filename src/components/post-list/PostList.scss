@import '../../variables';
.PostListContainer {
    display: flex;
    flex-direction: column;
    width: $POST_LIST_WIDTH;
    .MenuButton {
        display: none;
        top: -8px;
    }
    .PostListBody {
        flex: 1;
        display: flex;
        z-index: 1;
        background: #f5f2f2;
        flex-direction: column;
        height: 100%;
        width: unset;
        transform: translate3d(0%,0,0);
        top: 0;
    }
    .AnimationOpen {
        animation-duration: 1s !important;
    }
    .PostListOptions {
        display: flex;
        flex-direction: row;
        padding: 5px;
        .Option {
            flex: 1;
            
            padding: 5px 8px;
            &.PageSizeDropdown {
                margin-right: 4px;
                max-width: 100px;
                border: solid 0.4px black;
                box-shadow: 0px 3px 1px -2px rgb(0 0 0 / 20%), 0px 2px 2px 0px rgb(0 0 0 / 14%), 0px 1px 5px 0px rgb(0 0 0 / 12%);
                .PageSizeDropdownInput {
                    padding-left: 10px;
                }
            }
            &.ReloadButton {
                font-size: 12px !important;
                margin-left: 4px;
                margin-right: 4px;
            }
            &.DismissAllButton {
                margin-left: 4px;
                font-size: 12px !important;
                margin-right: 4px;
            }
            &.SavedPostsButton {
                font-size: 12px !important;
                margin-left: 4px;
                &.SavedButtonFocus { 
                    background-color: antiquewhite;
                }
            }
        }
    }
    .PostListInnerContainer {
        display: flex;
        flex: 1;
        flex-direction: column;
        margin-top: 5px;
        margin-bottom: 5px;
        padding-top: 20px;
        padding-bottom: 20px;
        padding-right: 20px;
        overflow-y: auto;
        border-right: groove 6px lightgrey;
    }
    @media screen and (max-width: $MOBILE_WIDTH_LIMIT) {
        width: unset;
        .PostListBody {
            width: $POST_LIST_WIDTH;
            position: fixed;
            transform: translate3d(-100%,0,0);
            height: calc(100% - 50px);
            top: 50px;
            border: solid 0.35px lightgrey;
        }
        .MenuButton {
            display: block;
        }
    }
}
