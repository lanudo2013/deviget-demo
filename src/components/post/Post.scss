.PostContainer {
    display: flex;
    flex-direction: column;
    margin-left: 10px;
    border-radius: 20px;
    margin-right: 10px;
    margin-top: 3px;
    padding: 20px;
    margin-bottom: 10px;
    box-shadow: 2px 1px 3px 2px rgb(191 191 191 / 42%);
    animation-duration: 0.8s !important;

    .Header {
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        .Author {
            font-size: 20px;
            font-weight: bold;
            line-height: 10px;
        }
    }
    .Body {
        min-height: 90px;
        display: flex;
        margin-top: 5px;
        align-items: center;
        border: solid 0px black;
        .Thumbnail {
        }
        .Title {
            vertical-align: center;
            margin-left: 10px;
            align-self: center;
        }
    }
    .Footer {
        margin-top: 7px;
        display: flex;
        justify-content: space-between;
        .DismissButton {
            padding-top: 1px;
            padding-bottom: 1px;
        }
        .CommentsNum {
            color: rgb(165, 165, 41);
            font-size: 16px;
            font-weight: bold;
        }
    }
}