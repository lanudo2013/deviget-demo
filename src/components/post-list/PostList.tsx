import { Button, IconButton, InputBase, MenuItem, Select } from '@material-ui/core';
import ClearAllIcon from '@material-ui/icons/ClearAll';
import MenuIcon from '@material-ui/icons/Menu';
import RefreshIcon from '@material-ui/icons/Refresh';
import { debounce } from 'lodash';
import React, { useEffect, useRef, useState } from 'react';
import { useDispatch, useSelector } from "react-redux";
import { AppState } from "../../classes/interfaces/appstate";
import { Post } from "../../classes/interfaces/post";
import { PostService } from "../../services/PostService";
import { requestPosts, retrieveDismissedPosts, retrieveReadPosts, saveDismissPost, saveDismissPosts, saveReadPost, selectPost, updateDoneDimissData, updateRequestPosts } from "../../state/actions";
import { PostRef, PostUI } from "../post/PostUI";
import { Util } from '../util';
import { Constants } from '../../constants';
import './PostList.scss';

interface PostListProps {
}
export const PostList = function(props: PostListProps) {

  const postService = React.useMemo(() => PostService.getInstance(), []);
  const [pageSize, setPageSize] = useState<number>(postService.getPageSize() || Constants.POST_LIST_PAGE_SIZES[0]);
  const [ready, setReady] = useState<boolean>(false);
  const [slideAnimating, setSlideAnimating] = useState<boolean>(false);
  const [menuAnimating, setMenuAnimating] = useState<boolean>(false);
  const [menuOpened, setMenuOpened] = useState<boolean | undefined>(undefined);
  const [windowDimensions, setWindowDimensions] = useState(Util.getWindowDimensions());

  useEffect(() => {
    setMenuOpened(true);
    function handleResize() {
      setMenuOpened(undefined);
      setWindowDimensions(Util.getWindowDimensions());
    }

    window.addEventListener('resize', handleResize);
    return () => window.removeEventListener('resize', handleResize);
  }, []);


  const dispatch = useDispatch();
  const initFetch = React.useCallback((oSize: number) => {
    postRefsMap.current = {};
    dispatch(updateRequestPosts([]));
    dispatch(requestPosts(oSize, true));
    dispatch(retrieveReadPosts());
    dispatch(retrieveDismissedPosts()); 
  }, [dispatch]);

  useEffect(() => {
    if (!ready) {
      postService.init().then(() => {
          setReady(true);
          initFetch(pageSize);
      }).catch(() => {
          setReady(true);
          initFetch(pageSize);
      });
    }
  }, [postService, dispatch]); 
   
    
  const posts = useSelector((state: AppState) => state.posts);
  const fetchingPosts = useSelector((state: AppState) => state.fetchingPosts);
  const dismissData = useSelector((state: AppState) => state.dismissData);
  const postRefsMap = useRef<{[key: string]: PostRef}>({});

  const runFadeOutAnimation = React.useCallback((postToDismiss: PostRef | undefined) => {
    if (postToDismiss) {
      postToDismiss.fadeOut().then(() => {
        dispatch(updateRequestPosts(posts.filter(x => x.id !== postToDismiss.id)));
        dispatch(updateDoneDimissData(undefined));
        delete postRefsMap.current[postToDismiss.id];
      }).catch(() => {});
    }
  }, [posts, dispatch]);

  const runSlideOutAnimation =React.useCallback(() => {
      let to = 0;
      const promises: Promise<any>[] = [];
      setSlideAnimating(true);
      posts.forEach((x, i) => {
        const postRef = postRefsMap.current[x.id];
        to = Math.min(600, i * 25);
        promises.push(postRef.slideOut(to));
      });

      Promise.all(promises).then(() => {
        dispatch(updateDoneDimissData(undefined));
        dispatch(updateRequestPosts([]));
        postRefsMap.current = {};
        setSlideAnimating(false);
      }).catch(() => {
        setSlideAnimating(false);
      });
  }, [posts, dispatch]);



  useEffect(() => {
    if (posts.length === Object.keys(postRefsMap.current).length && dismissData) {
        if (dismissData.type === 'fadeOut') {
          const postToDismiss = postRefsMap.current[dismissData.id || ''];
          runFadeOutAnimation(postToDismiss);
        } else if (dismissData.type === 'slideOut') {
          runSlideOutAnimation();
        }
    }
  }, [dismissData, runSlideOutAnimation, runFadeOutAnimation, posts]);

  const onDismissed = React.useCallback((p: Post) => {
      dispatch(saveDismissPost(p.id));
  }, [dispatch]);
  const onPress = React.useCallback((p: Post) => {
      dispatch(saveReadPost(p.id));
      dispatch(selectPost(p));
      if (windowDimensions.width < Constants.MOBILE_WIDHT_LIMIT_PX) {
        setMenuOpened(false);
      }
      
  }, [dispatch, windowDimensions]);

    const onScroll = React.useCallback(
      debounce((e: any) => {
        const bottomReached = e.target.scrollHeight - e.target.scrollTop <= e.target.clientHeight * 1.1 && e.target.scrollTop > 0;
        if (bottomReached && !fetchingPosts && !slideAnimating) {
            dispatch(requestPosts(pageSize));
        }
      }, 350), 
    [dispatch, pageSize, fetchingPosts, slideAnimating]);
  
    const postsEl = React.useMemo(() => {
      return posts.map(x => <PostUI key={x.id} post={x} currentDate={new Date()} onPressDismiss={onDismissed} onPress={onPress} ref={r => r && (postRefsMap.current[r.id]=r)}></PostUI>);
    }, [posts]);

    const handleChange = React.useCallback((ev: any) => {
        if (ev && ev.target && ev.target.value) {
            setPageSize(ev.target.value);
            postService.savePageSize(ev.target.value);
            initFetch(ev.target.value);
         }
        
    }, [initFetch, postService]);

    const refreshPress = React.useCallback((ev: any) => {
        initFetch(pageSize);
    }, [initFetch, pageSize]);

    const dismissAllPress = React.useCallback((ev: any) => {
      if (!slideAnimating) {
        const ids = posts.map(x => x.id);
        dispatch(saveDismissPosts(ids));
      }
  }, [slideAnimating, posts, dispatch]);

    const onAnimSliderEnd = React.useCallback(() => {
      setMenuAnimating(false);
    }, []);

    const openMenu = React.useCallback(() => {
      if (!menuAnimating) {
        setMenuOpened(x => !x);
        setMenuAnimating(true);
      }
    }, [menuAnimating]);

    const pageSizes = React.useMemo(() => {
      return Constants.POST_LIST_PAGE_SIZES.map(x => ( <MenuItem value={x}>{x}</MenuItem>));
    }, []);

    return <div className="PostListContainer">
      <IconButton aria-label="delete" className={'MenuButton '} onClick={openMenu}>
          <MenuIcon fontSize="large" />
      </IconButton>
      <div className={'PostListBody ' +  (windowDimensions.width < Constants.MOBILE_WIDHT_LIMIT_PX && menuOpened !== undefined ? (menuOpened ? 'animate__animated animate__slideInLeft' : 'animate__animated animate__slideOutLeft') : '')} onAnimationEnd={onAnimSliderEnd}>
        <div className="PostListOptions">
          <Select
              value={pageSize}
              className="Option PageSizeDropdown"
              onChange={handleChange}
              input={<InputBase placeholder={Constants.APP_MESSAGES.PAGE_SIZE_PLC} style={{paddingLeft: '12px'}}/>}
            >
              <MenuItem value="">
                <em>{Constants.APP_MESSAGES.PAGE_SIZE_PLC}</em>
              </MenuItem>
              {pageSizes}
            </Select>
            <Button
              variant="contained"
              className="Option ReloadButton"
              onClick={refreshPress}
              endIcon={<RefreshIcon />}
            >
              {Constants.APP_MESSAGES.REFRESH_BUTTON}
            </Button> 
            <Button
              variant="contained"
              className="Option DismissButton"
              onClick={dismissAllPress}
              endIcon={<ClearAllIcon />}
            >
              {Constants.APP_MESSAGES.DISMISS_ALL_BUTTON}
            </Button>

        </div>
        <div className="PostListInnerContainer" style={!postsEl.length ? {border: 'none'} : {}} onScroll={onScroll}>
            {postsEl}
        </div>
      </div>
    </div>
    ;
};