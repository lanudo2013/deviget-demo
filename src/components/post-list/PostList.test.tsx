import { PostType } from '../../classes/enums/post-type';
import { Post } from '../../classes/interfaces/post';
import { mount, ReactWrapper, shallow } from 'enzyme';
import moment from 'moment';
import React, { useRef } from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { PostList } from './PostList';
import { Button, MenuItem, Select } from '@material-ui/core';
import { PostService } from '../../services/PostService';
import { Constants } from '../../constants';
import { AppState } from '../../classes/interfaces/appstate';
import { buildMockedPromise, createMockedPostServiceObj } from '../../mocks/post-service';
import { PostUI } from '../post/PostUI';
import { createStore, Store } from 'redux';
import rootReducer from '../../state/rootReducer';

const globalStoreRef = { store: createStore(rootReducer) };
const mockDispatch = jest.fn().mockImplementation((val) => {
    if (typeof val === 'function') {
        return val(mockDispatch, globalStoreRef.store.getState);
    }
    return globalStoreRef.store.dispatch(val);
});
jest.mock('react-redux', () => {
    return {
        useSelector: (fn: (x: AppState) => any) => fn(globalStoreRef.store.getState()),
        useDispatch: () => mockDispatch
    };
});
jest.mock('../../services/PostService', () => {
    let instance: any = null;
    return {
        PostService: {
            getInstance: jest.fn().mockImplementation(() => {
                if (instance) {
                    return instance;
                }
                instance = createMockedPostServiceObj();
                return instance;
            })
        }
    };
});

describe('Post component', () => {
    let root: ReactWrapper;
    let listP: Post[];
    const currentDate = new Date();

    function createPosts(): Post[] {
        return [
            {
                author: 'author1',
                createdTime: moment(currentDate)
                    .add(-4, 'days')
                    .add(-6, 'hours')
                    .add(+new Date().getTimezoneOffset(), 'minutes')
                    .toDate(),
                numberOfComments: 1230,
                id: 'p1',
                postType: PostType.IMAGE,
                thumbnailUrl: '/p1-thumb.jpg',
                title: 'My new post 1',
                subreddit: 'Sub',
                postUrl: '/p1.jpg',
                postHtml: null
            },
            {
                author: 'author1',
                createdTime: moment(currentDate)
                    .add(-10, 'days')
                    .add(+new Date().getTimezoneOffset(), 'minutes')
                    .toDate(),
                numberOfComments: 26,
                id: 'p2',
                postType: PostType.VIDEO,
                thumbnailUrl: '/p2-thumb.jpg',
                title: 'My new post 2',
                subreddit: 'Sub2',
                postUrl: '/p2.jpg',
                postHtml: null,
                videoData: {
                    url: 'http://myurl.com/video.mpg',
                    width: 200,
                    height: 200
                }
            },
            {
                author: 'author2',
                createdTime: moment(currentDate)
                    .add(-2, 'minutes')
                    .add(+new Date().getTimezoneOffset(), 'minutes')
                    .toDate(),
                numberOfComments: 900,
                id: 'p3',
                postType: PostType.LINK,
                thumbnailUrl: undefined,
                title: 'My new post 3',
                subreddit: 'Sub3',
                postUrl: 'http://www.google.com',
                postHtml: null
            }
        ];
    }
    listP = createPosts();
    const postService = PostService.getInstance();
    jest.spyOn(postService, 'getPosts').mockImplementation(
        (l: number, reset: boolean): Promise<Post[]> => {
            return buildMockedPromise(listP);
        }
    );

    function doBasicChecks(node: any) {
        const tree = renderer.create(node).toJSON();
        expect(tree).toMatchSnapshot();

        let domElement: ReactWrapper = root.update().find(PostList).find('div.PostListContainer');
        expect(domElement.length).toEqual(1);
        expect(domElement.get(0)).toBeTruthy();
        domElement = root.find('div.PostListContainer > div.PostListBody');
        expect(domElement.length).toEqual(1);
        expect(domElement.get(0)).toBeTruthy();
        domElement = root.find('div.PostListBody > div.PostListInnerContainer');
        expect(domElement.length).toEqual(1);
        expect(domElement.get(0)).toBeTruthy();
        domElement = root.find('div.PostListBody > div.PostListOptions');
        expect(domElement.length).toEqual(1);
        expect(domElement.get(0)).toBeTruthy();
        expect(domElement.children().length).toEqual(4);
        domElement = root.find('div.PostListOptions').find(Select);
        expect(domElement.length).toEqual(1);
        expect(domElement.get(0)).toBeTruthy();
    }

    beforeEach((dn) => {
        listP = createPosts();
        globalStoreRef.store = createStore(rootReducer);
        dn();
    });

    afterEach(() => {
        jest.clearAllMocks();
        jest.resetModules();
    });

    test('Init component with some posts', () => {
        const p = listP;
        root = mount(<PostList />, {
            wrappingComponent: Provider,
            wrappingComponentProps: {
                store: globalStoreRef.store
            }
        });
        doBasicChecks(<PostList />);
        expect(postService.init).toHaveBeenCalled();
        const postsCalls = mockDispatch.mock.calls.filter(
            (x: any) => x[0] && x[0].type && x[0].type === Constants.REDUX_ACTIONS.UPDATE_POST_LIST
        );
        expect(postsCalls.length).toEqual(2);
        expect(postsCalls[0][0].payload).toEqual([]);
        expect(postsCalls[1][0].payload).toEqual(p);
        expect(postService.getPosts).toHaveBeenCalled();
        expect(postService.getReadPosts).toHaveBeenCalled();
        expect(postService.getDismissedPosts).toHaveBeenCalled();

        const domElements = root.update().find(PostUI);
        expect(domElements.length).toEqual(3);
    });

    test('Select one post', (dn) => {
        const p = listP;
        root = mount(<PostList />, {
            wrappingComponent: Provider,
            wrappingComponentProps: {
                store: globalStoreRef.store
            }
        });

        const domElements = root.update().find(PostUI);
        expect(domElements.length).toEqual(3);
        const postUi = domElements.at(0).find('div.PostContainer');
        expect(postUi.get(0)).toBeTruthy();
        postUi.simulate('click');
        expect(postService.saveReadPost).toHaveBeenCalled();
        postService.getReadPosts().then((rs) => {
            expect(rs.indexOf(p[0].id) >= 0).toBeTruthy();
            expect(globalStoreRef.store.getState().selectedPost).toHaveProperty('id', p[0].id);
            expect(globalStoreRef.store.getState().postsRead).toContain(p[0].id);
            dn();
        });
    });

    test('Save dismiss one post', (dn) => {
        const p = listP;
        root = mount(<PostList />, {
            wrappingComponent: Provider,
            wrappingComponentProps: {
                store: globalStoreRef.store
            }
        });

        const domElements = root.update().find(PostUI);
        expect(domElements.length).toEqual(3);
        const postUi = domElements.at(0).find('div.PostContainer');
        expect(postUi.get(0)).toBeTruthy();
        postUi.find(Button).find('button.DismissButton').simulate('click');
        expect(postService.saveDismissPost).toHaveBeenCalled();
        postService.getDismissedPosts().then((rs) => {
            expect(rs.indexOf(p[0].id) >= 0).toBeTruthy();
            expect(globalStoreRef.store.getState().dismissData).toEqual({ id: p[0].id, type: 'fadeOut' });
            dn();
        });
    });

    test('Select and dismiss same post', (dn) => {
        const p = listP;
        root = mount(<PostList />, {
            wrappingComponent: Provider,
            wrappingComponentProps: {
                store: globalStoreRef.store
            }
        });

        const domElements = root.update().find(PostUI);
        expect(domElements.length).toEqual(3);
        const postUi = domElements.at(0).find('div.PostContainer');
        expect(postUi.get(0)).toBeTruthy();
        postUi.simulate('click');
        expect(postUi.get(0)).toBeTruthy();
        postUi.find(Button).find('button.DismissButton').simulate('click');
        expect(postService.saveDismissPost).toHaveBeenCalled();
        setTimeout(() => {
            postService.getDismissedPosts().then((rs) => {
                expect(rs.indexOf(p[0].id) >= 0).toBeTruthy();
                expect(globalStoreRef.store.getState().selectedPost).toBeFalsy();
                dn();
            });
        });
    });

    test('Dismiss all posts', (dn) => {
        const p = listP;
        root = mount(<PostList />, {
            wrappingComponent: Provider,
            wrappingComponentProps: {
                store: globalStoreRef.store
            }
        });

        const dismissAllButton = root.update().find('button.DismissAllButton');
        expect(dismissAllButton.get(0)).toBeTruthy();
        dismissAllButton.simulate('click');
        expect(postService.saveDismissPosts).toHaveBeenCalled();
        postService.getDismissedPosts().then((rs) => {
            expect(rs.length).toEqual(3);
            expect(rs.indexOf(p[0].id) >= 0).toBeTruthy();
            expect(rs.indexOf(p[1].id) >= 0).toBeTruthy();
            expect(rs.indexOf(p[2].id) >= 0).toBeTruthy();
            expect(globalStoreRef.store.getState().dismissData).toEqual({ type: 'slideOut' });
            dn();
        });
    });

    test('Scroll to bottom and call reload posts request', (dn) => {
        root = mount(<PostList />, {
            wrappingComponent: Provider,
            wrappingComponentProps: {
                store: globalStoreRef.store
            }
        });

        const postInnerContainer = root.update().find('div.PostListInnerContainer');
        postInnerContainer.simulate('scroll', { target: { scrollHeight: 100, scrollTop: 99, clientHeight: 100 } });
        setTimeout(() => {
            expect(postService.getPosts).toHaveBeenCalledTimes(2);
            dn();
        }, 500);
    });

    test('Save two first posts and show savedx', (dn) => {
        const p = listP;
        root = mount(<PostList />, {
            wrappingComponent: Provider,
            wrappingComponentProps: {
                store: globalStoreRef.store
            }
        });

        const domElements = root.update().find(PostUI);
        expect(domElements.length).toEqual(3);
        let postUi = domElements.at(0).find('button.SaveButton');
        expect(postUi.get(0)).toBeTruthy();
        postUi.simulate('click');
        postUi = domElements.at(2).find('button.SaveButton');
        expect(postUi.get(0)).toBeTruthy();
        postUi.simulate('click');
        expect(postService.savePost).toHaveBeenCalledTimes(2);

        setTimeout(() => {
            postService.getSavedPosts().then((rs) => {
                expect(rs).toEqual([p[0], p[2]]);
                expect(globalStoreRef.store.getState().posts).toEqual([p[0], p[1], p[2]]);
                expect(globalStoreRef.store.getState().savedPosts).toEqual([p[0], p[2]]);
                const postButton = root.update().find('div.PostListOptions').find('button.SavedPostsButton');
                expect(postButton.get(0)).toBeTruthy();
                postButton.simulate('click');
                setTimeout(() => {
                    expect(globalStoreRef.store.getState().showSaved).toEqual(true);
                    const domElements1 = root.update().find(PostUI);
                    expect(domElements1.length).toEqual(2);
                    const foundPosts: Post[] = [];
                    domElements1.forEach((el, idx) => {
                        const p = el.getElement().props.post;
                        foundPosts.push(p);
                    });
                    expect(foundPosts).toEqual([p[0], p[2]]);
                    dn();
                }, 1000);
            });
        });
    });
});
