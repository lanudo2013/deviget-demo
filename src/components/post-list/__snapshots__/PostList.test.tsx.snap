// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post component Init component with some posts 1`] = `
<div
  className="PostListContainer"
>
  <button
    aria-label="delete"
    className="MuiButtonBase-root MuiIconButton-root MenuButton "
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
        />
      </svg>
    </span>
  </button>
  <div
    className="PostListBody "
    onAnimationEnd={[Function]}
  >
    <div
      className="PostListOptions"
    >
      <div
        className="MuiInputBase-root Option PageSizeDropdown"
        onClick={[Function]}
        style={
          Object {
            "paddingLeft": "12px",
          }
        }
      >
        <div
          aria-haspopup="listbox"
          className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          25
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput"
          onAnimationStart={[Function]}
          onChange={[Function]}
          placeholder="Page Size"
          tabIndex={-1}
          value={25}
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained Option ReloadButton"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Refresh
          <span
            className="MuiButton-endIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
              />
            </svg>
          </span>
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained Option DismissButton"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Dismiss All
          <span
            className="MuiButton-endIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M5 13h14v-2H5v2zm-2 4h14v-2H3v2zM7 7v2h14V7H7z"
              />
            </svg>
          </span>
        </span>
      </button>
    </div>
    <div
      className="PostListInnerContainer"
      onScroll={[Function]}
      style={Object {}}
    >
      <div
        className="PostContainer "
        onAnimationEnd={[Function]}
        onClick={[Function]}
        style={Object {}}
      >
        <div
          className="Header"
        >
          <div
            className="Header-left"
          >
            <span
              aria-hidden={true}
              className="material-icons MuiIcon-root ReadIcon"
            >
              fiber_manual_record
            </span>
            <span
              className="Author"
              style={null}
            >
              author1
            </span>
          </div>
          <span
            className="CreatedAt"
          >
            4 days, 6 hours ago
          </span>
        </div>
        <div
          className="Body"
        >
          <img
            className="Thumbnail"
            height={100}
            src="/p1-thumb.jpg"
            width={100}
          />
          <span
            className="Title"
          >
            My new post 1
          </span>
        </div>
        <div
          className="Footer"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeSmall MuiButton-sizeSmall"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "fontSize": "12px",
                "paddingBottom": "1px",
                "paddingTop": "1px",
              }
            }
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              <span
                className="MuiButton-startIcon MuiButton-iconSizeSmall"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm3.59-13L12 10.59 8.41 7 7 8.41 10.59 12 7 15.59 8.41 17 12 13.41 15.59 17 17 15.59 13.41 12 17 8.41z"
                  />
                </svg>
              </span>
              Dismiss Post
            </span>
          </button>
          <span
            className="CommentsNum"
          >
            1,230 comment/s
          </span>
        </div>
      </div>
      <div
        className="PostContainer "
        onAnimationEnd={[Function]}
        onClick={[Function]}
        style={Object {}}
      >
        <div
          className="Header"
        >
          <div
            className="Header-left"
          >
            <span
              aria-hidden={true}
              className="material-icons MuiIcon-root ReadIcon"
            >
              fiber_manual_record
            </span>
            <span
              className="Author"
              style={null}
            >
              author1
            </span>
          </div>
          <span
            className="CreatedAt"
          >
            10 days ago
          </span>
        </div>
        <div
          className="Body"
        >
          <img
            className="Thumbnail"
            height={100}
            src="/p2-thumb.jpg"
            width={100}
          />
          <span
            className="Title"
          >
            My new post 2
          </span>
        </div>
        <div
          className="Footer"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeSmall MuiButton-sizeSmall"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "fontSize": "12px",
                "paddingBottom": "1px",
                "paddingTop": "1px",
              }
            }
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              <span
                className="MuiButton-startIcon MuiButton-iconSizeSmall"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm3.59-13L12 10.59 8.41 7 7 8.41 10.59 12 7 15.59 8.41 17 12 13.41 15.59 17 17 15.59 13.41 12 17 8.41z"
                  />
                </svg>
              </span>
              Dismiss Post
            </span>
          </button>
          <span
            className="CommentsNum"
          >
            26 comment/s
          </span>
        </div>
      </div>
      <div
        className="PostContainer "
        onAnimationEnd={[Function]}
        onClick={[Function]}
        style={Object {}}
      >
        <div
          className="Header"
        >
          <div
            className="Header-left"
          >
            <span
              aria-hidden={true}
              className="material-icons MuiIcon-root ReadIcon"
            >
              fiber_manual_record
            </span>
            <span
              className="Author"
              style={null}
            >
              author2
            </span>
          </div>
          <span
            className="CreatedAt"
          >
            2 minutes ago
          </span>
        </div>
        <div
          className="Body"
        >
          <span
            className="Title"
          >
            My new post 3
          </span>
        </div>
        <div
          className="Footer"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeSmall MuiButton-sizeSmall"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "fontSize": "12px",
                "paddingBottom": "1px",
                "paddingTop": "1px",
              }
            }
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              <span
                className="MuiButton-startIcon MuiButton-iconSizeSmall"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm3.59-13L12 10.59 8.41 7 7 8.41 10.59 12 7 15.59 8.41 17 12 13.41 15.59 17 17 15.59 13.41 12 17 8.41z"
                  />
                </svg>
              </span>
              Dismiss Post
            </span>
          </button>
          <span
            className="CommentsNum"
          >
            900 comment/s
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
